File: backend/src/main/java/com/stjoseph/assessmentsystem/model/Assessment.java
package com.stjoseph.assessmentsystem.model;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
import java.util.List;

@Document(collection = "assessments")
public class Assessment {
    @Id
    private String id;
    
    private String title;
    private String description;
    private String createdBy;
    private List<String> assignedTo;
    private List<Question> questions;
    private LocalDateTime startTime;
    private LocalDateTime endTime;
    private LocalDateTime createdAt;
    private AssessmentType type;
    private String domain;
    private String difficulty;
    private int totalMarks;
    private int duration; // in minutes
    
    public enum AssessmentType {
        AI_GENERATED, CLASS_ASSESSMENT
    }
    
    public static class Question {
        private String question;
        private List<String> options;
        private int correctAnswer;
        private String explanation;
        
        // Getters and Setters
        public String getQuestion() { return question; }
        public void setQuestion(String question) { this.question = question; }
        
        public List<String> getOptions() { return options; }
        public void setOptions(List<String> options) { this.options = options; }
        
        public int getCorrectAnswer() { return correctAnswer; }
        public void setCorrectAnswer(int correctAnswer) { this.correctAnswer = correctAnswer; }
        
        public String getExplanation() { return explanation; }
        public void setExplanation(String explanation) { this.explanation = explanation; }
    }
    
    public Assessment() {
        this.createdAt = LocalDateTime.now();
    }
    
    // Getters and Setters
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }
    
    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }
    
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    
    public String getCreatedBy() { return createdBy; }
    public void setCreatedBy(String createdBy) { this.createdBy = createdBy; }
    
    public List<String> getAssignedTo() { return assignedTo; }
    public void setAssignedTo(List<String> assignedTo) { this.assignedTo = assignedTo; }
    
    public List<Question> getQuestions() { return questions; }
    public void setQuestions(List<Question> questions) { this.questions = questions; }
    
    public LocalDateTime getStartTime() { return startTime; }
    public void setStartTime(LocalDateTime startTime) { this.startTime = startTime; }
    
    public LocalDateTime getEndTime() { return endTime; }
    public void setEndTime(LocalDateTime endTime) { this.endTime = endTime; }
    
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    
    public AssessmentType getType() { return type; }
    public void setType(AssessmentType type) { this.type = type; }
    
    public String getDomain() { return domain; }
    public void setDomain(String domain) { this.domain = domain; }
    
    public String getDifficulty() { return difficulty; }
    public void setDifficulty(String difficulty) { this.difficulty = difficulty; }
    
    public int getTotalMarks() { return totalMarks; }
    public void setTotalMarks(int totalMarks) { this.totalMarks = totalMarks; }
    
    public int getDuration() { return duration; }
    public void setDuration(int duration) { this.duration = duration; }
}

File: lib/models/assessment.dart
class Assessment {
  final String id;
  final String title;
  final String description;
  final String createdBy;
  final List<String> assignedTo;
  final List<Question> questions;
  final DateTime startTime;
  final DateTime endTime;
  final DateTime createdAt;
  final AssessmentType type;
  final String? domain;
  final String? difficulty;
  final int totalMarks;
  final int duration;

  const Assessment({
    required this.id,
    required this.title,
    required this.description,
    required this.createdBy,
    required this.assignedTo,
    required this.questions,
    required this.startTime,
    required this.endTime,
    required this.createdAt,
    required this.type,
    this.domain,
    this.difficulty,
    required this.totalMarks,
    required this.duration,
  });

  factory Assessment.fromJson(Map<String, dynamic> json) {
    return Assessment(
      id: json['id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      createdBy: json['createdBy'] ?? '',
      assignedTo: json['assignedTo'] != null ? List<String>.from(json['assignedTo']) : [],
      questions: json['questions'] != null 
          ? (json['questions'] as List).map((q) => Question.fromJson(q)).toList()
          : [],
      startTime: DateTime.parse(json['startTime'] ?? DateTime.now().toIso8601String()),
      endTime: DateTime.parse(json['endTime'] ?? DateTime.now().toIso8601String()),
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      type: AssessmentType.values.firstWhere(
        (e) => e.name.toUpperCase() == (json['type'] ?? 'AI_GENERATED').toString().toUpperCase(),
        orElse: () => AssessmentType.AI_GENERATED,
      ),
      domain: json['domain'],
      difficulty: json['difficulty'],
      totalMarks: json['totalMarks'] ?? 0,
      duration: json['duration'] ?? 60,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'createdBy': createdBy,
      'assignedTo': assignedTo,
      'questions': questions.map((q) => q.toJson()).toList(),
      'startTime': startTime.toIso8601String(),
      'endTime': endTime.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'type': type.name.toUpperCase(),
      'domain': domain,
      'difficulty': difficulty,
      'totalMarks': totalMarks,
      'duration': duration,
    };
  }
}

class Question {
  final String question;
  final List<String> options;
  final int correctAnswer;
  final String explanation;

  const Question({
    required this.question,
    required this.options,
    required this.correctAnswer,
    required this.explanation,
  });

  factory Question.fromJson(Map<String, dynamic> json) {
    return Question(
      question: json['question'] ?? '',
      options: json['options'] != null ? List<String>.from(json['options']) : [],
      correctAnswer: json['correctAnswer'] ?? 0,
      explanation: json['explanation'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'question': question,
      'options': options,
      'correctAnswer': correctAnswer,
      'explanation': explanation,
    };
  }
}

enum AssessmentType {
  AI_GENERATED,
  CLASS_ASSESSMENT,
}

extension AssessmentTypeExtension on AssessmentType {
  String get displayName {
    switch (this) {
      case AssessmentType.AI_GENERATED:
        return 'AI Generated';
      case AssessmentType.CLASS_ASSESSMENT:
        return 'Class Assessment';
    }
  }
}